weblate/vcs/git.py

```diff
@@ -1468,3 +1468,120 @@ def create_pull_request(
             raise RepositoryException(0, f"Pull request failed: {error_message}")
+
+
+class BitbucketServerRepository(GitMergeRequestBase):
+    name = gettext_lazy("Bitbucket Server pull request")
+    identifier = "bitbucketserver"
+    _version = None
+    API_TEMPLATE = "https://{host}/rest/api/1.0/projects/{owner}/repos/{slug}"
+    bb_fork: Dict = {}
+
+    def get_headers(self, credentials: Dict):
+        headers = super().get_headers(credentials)
+        headers["Authorization"] = f"Bearer {credentials['token']}"
+        return headers
+
+    def create_fork(self, credentials: Dict):
+        fork_url, owner, slug = self.get_api_url()
+        bb_fork, error_message = self.request("post", credentials, fork_url, json={})
+        self.bb_fork = bb_fork
+
+        # If fork already exists, get forks from origin and find the user's fork.
+        # Since Bitbucket uses projectKey which can be any string(that is not
+        # at all related to its name) we need to compare user's forks against
+        # remote.
+        if "This repository URL is already taken." in error_message:
+            page = 0
+            self.bb_fork = {}
+            while True:
+                forks_url = f'{credentials["url"]}/forks?limit=1000&start={page}'
+                forks, error_message = self.request("get", credentials, forks_url)
+                if "values" in forks:
+                    for f in forks["values"]:
+                        fork_slug = f["origin"]["slug"]
+                        fork_project_key = f["origin"]["project"]["key"]
+                        if (
+                            fork_slug == slug
+                            and fork_project_key.upper() == owner.upper()
+                        ):
+                            self.bb_fork = f
+                            break
+
+                if self.bb_fork or forks["isLastPage"] or not forks["values"]:
+                    break
+
+                page += 1
+
+        remote_url = None
+        if "links" in self.bb_fork:
+            clone_urls = self.bb_fork["links"]["clone"]
+            for link in clone_urls:
+                if link["name"] == "ssh":
+                    remote_url = link["href"]
+
+        if not remote_url:
+            raise RepositoryException(
+                0, f"Failed to create or find bitbucket fork. {error_message}"
+            )
+
+        self.configure_fork_remote(remote_url, credentials["username"])
+
+    def get_default_reviewers(self, credentials: Dict, fork_branch: str):
+        target_repo, error_message = self.request(
+            "get", credentials, credentials["url"]
+        )
+        if "id" not in target_repo:
+            return []
+
+        url = credentials["url"].replace(
+            "/rest/api/1.0/", "/rest/default-reviewers/1.0/"
+        )
+        url = f'{url}/reviewers?targetRepoId={target_repo["id"]}&sourceRepoId={self.bb_fork["id"]}&targetRefId={self.branch}&sourceRefId={fork_branch}'
```

Comment:
Please pass the parameters to `request` as `params` dict instead of building the long URL here.

Reply:
fixed in amend.